cluster/reconcile.go

```diff
@@ -0,0 +1,134 @@
+package cluster
+
+import (
+	"fmt"
+
+	"github.com/rancher/rke/hosts"
+	"github.com/rancher/rke/k8s"
+	"github.com/rancher/rke/services"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ssh"
+	"k8s.io/client-go/kubernetes"
+)
+
+func ReconcileCluster(kubeCluster, currentCluster *Cluster) error {
+	logrus.Infof("[reconcile] Reconciling cluster state")
+	if currentCluster == nil {
+		logrus.Infof("[reconcile] This is newly generated cluster")
+
+		return nil
+	}
+	// to handle if current local admin is down and we need to use new cp from the list
+	if !isLocalConfigWorking(kubeCluster.LocalKubeConfigPath) {
+		if err := rebuildLocalAdminConfig(kubeCluster); err != nil {
+			return err
+		}
+	}
+
+	kubeClient, err := k8s.NewClient(kubeCluster.LocalKubeConfigPath)
+	if err != nil {
+		return fmt.Errorf("Failed to initialize new kubernetes client: %v", err)
+	}
+
+	key, _ := checkEncryptedKey(kubeCluster.SSHKeyPath)
+
+	if err := reconcileWorker(currentCluster, kubeCluster, key, kubeClient); err != nil {
+		return err
+	}
+
+	if err := reconcileControl(currentCluster, kubeCluster, key, kubeClient); err != nil {
+		return err
+	}
+	logrus.Infof("[reconcile] Reconciled cluster state successfully")
+	return nil
+}
+
+func reconcileWorker(currentCluster, kubeCluster *Cluster, key ssh.Signer, kubeClient *kubernetes.Clientset) error {
+	// worker deleted first to avoid issues when worker+controller on same host
+	logrus.Infof("[reconcile] Check worker hosts to be deleted")
+	wpToDelete := hosts.GetToDeleteHosts(currentCluster.WorkerHosts, kubeCluster.WorkerHosts)
+	for _, toDeleteHost := range wpToDelete {
+		if err := currentCluster.removeRoleFromNode(toDeleteHost, services.WorkerRole); err != nil {
+			return err
+		}
+		if err := hosts.DeleteNode(&toDeleteHost, kubeClient, services.ControlRole); err != nil {
+			return fmt.Errorf("Failed to delete worker node %s from cluster", toDeleteHost.Address)
+		}
+		// attempting to clean services/files on the host
+		if err := reconcileHost(toDeleteHost, key, true); err != nil {
+			logrus.Warnf("[reconcile] Couldn't clean up worker node [%s]: %v", toDeleteHost.Address, err)
+			continue
+		}
+	}
+	if err := currentCluster.InvertIndexHosts(); err != nil {
```

Comment:
I feel like this check belongs to the caller method. Also it is a bit concerning we are calling this check method after delete is already performed. I'd do it as a very first step.